R/dplyr_methods.R

```diff
@@ -275,12 +274,13 @@ mutate.SingleCellExperiment <- function(.data, ...) {
 rename.SingleCellExperiment <- function(.data, ...) {
 
     # Check that we are not modifying a key column
-    cols <- tidyselect::eval_select(expr(c(...)), colData(.data) %>% as.data.frame())
-
+    df <- as_tibble(.data)
```

Comment:
This would be, I believe more inefficient than getting colData directly, as it integrates also PC, UMAP etc..

Reply:
maybe could you help me understand the nature of this 3-line change? So I can have a bit more context.

Reply:
Yes, so I came across this while writing unit tests. Basically, if you try with the current version something like `rename(sce, new_name=.cell)`, it will pass without an informative warning, but then fails because `.cell` is not an actual column.

I think it works for other functions, but not here, because the syntax is `rename(..., new=old)`, so `cols` will correspond to the left-hand side names, not the ones being changed (right-hand side).

What made it clear to me was adding a `print(cols)` in line 279 and see what's happening vs. what happens with using `eval_rename()` instead (which returns indices)... Of course, would be great to find a way without using `as_tibble()`, but I couldn't come up with one :/ Though if we want to limit cost here, we could just do, e.g., `as_tibble(.data[, 1])`.

Reply:
Actually, I am not sure if maybe you head in mind doing both, i.e., not letting someone rename a column to `.cell`, but also not trying to rename `.cell`? I that case, we'd need a combination of both `eval_select()` and something else like the above, but more elegantly perhaps.

Reply:
I see. 

So .cell should be an immutable column. 

A solution can be 

- colData() |> as_tibble(rownames = EXISTING_INTERNAL_FUNCTION_THAT_ENCODES_CELL_COLNAME)
- rename
- drop .cell column so it's untouched
- warning ("tidySingleCellExperiment says: ...") If .cell is in the query (there could be multiple columns in the query)

Reply:
> Actually, I am not sure if maybe you head in mind doing both, i.e., not letting someone rename a column to `.cell`, but also not trying to rename `.cell`?

Pretty much this.